src/v/kafka/client/consumer.cc

```diff
@@ -475,22 +478,119 @@ ss::future<fetch_response> consumer::fetch(
       detail::reduce_fetch_response);
 }
 
+template<typename request_factory>
+ss::future<
+  typename std::invoke_result_t<request_factory>::api_type::response_type>
+consumer::reset_coordinator_and_retry_request(request_factory req) {
+    return find_coordinator_with_retry_and_mitigation(
+             _gate,
+             _config,
+             _brokers,
+             group_id(),
+             name(),
+             [this](std::exception_ptr ex) { return _external_mitigate(ex); })
+      .then(
+        [this, req{std::move(req)}](shared_broker_t new_coordinator) mutable {
+            _coordinator = new_coordinator;
+            // Calling req_res here will re-issue the request on the
+            // new coordinator
+            return req_res(std::move(req));
+        });
+}
+
+template<typename request_factory, typename response_t>
+ss::future<response_t>
+consumer::maybe_process_response_errors(request_factory req, response_t res) {
+    auto me = shared_from_this();
+    if constexpr (std::is_same_v<response_t, metadata_response>) {
```

Comment:
looks good but it's not clear what we are gaining from having all the constexpr checks. why not use normal overloading:

maybe_process_response_errors(request_factory req, metadata_response);
maybe_process_response_errors(request_factory req, offset_commit_response);
...

Reply:
Long answer incoming.

The consumer executes all requests with the `consumer::req_res` method except for `consumer::fetch` which is run on the leader for each topic-partition. Here is the definition of `consumer::req_res` (so you don't have to look through the diff):
```cpp
    template<typename RequestFactory>
    ss::future<
      typename std::invoke_result_t<RequestFactory>::api_type::response_type>
    req_res(RequestFactory req) {
        using api_t = typename std::invoke_result_t<RequestFactory>::api_type;
        using response_t = typename api_t::response_type;
        return ss::try_with_gate(_gate, [this, req{std::move(req)}]() mutable {
            auto r = req();
            kclog.debug(
              "Consumer: {}: {} req: {}, coordinator {}",
              *this,
              api_t::name,
              r,
              _coordinator->id());
            return _coordinator->dispatch(std::move(r))
              .then([this, req{std::move(req)}](response_t res) mutable {
                  kclog.debug(
                    "Consumer: {}: {} res: {}, coordinator {}",
                    *this,
                    api_t::name,
                    res,
                    _coordinator->id());
                  return maybe_process_response_errors(
                    std::move(req), std::move(res));
              });
        });
    }
```

Notice that the response type is deduced from the request factory. Therefore, I cannot pass a specific response type into `maybe_process_response_errors`. Of course I **could** check for response types in `req_res` but IMO that seems silly when I could do that in `mayber_process_response_errors` instead.

If that does not answer your question, then let's sync via video chat.

Reply:
No need for `constexpr` instead, do method overloading.